numpy/core/src/multiarray/arrayobject.c

```diff
@@ -534,6 +545,8 @@ array_dealloc(PyArrayObject *self)
     npy_free_cache_dim(fa->dimensions, 2 * fa->nd);
     Py_DECREF(fa->descr);
     Py_TYPE(self)->tp_free((PyObject *)self);
+
+    Py_TRASHCAN_SAFE_END(self)
```

Comment:
Hmmm, I am not sure whether the other objects may require something similar. OTOH: only arrays can contain arrays. All other objects should always contain something which has the `TRASHCAN_SAFE` mechanism at some point I guess.